app/models/potential_matcher/prompt_tag_type_info.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+# A class used for gathering information about the rough frequency of different
+# required tag types in all of the prompts in a collection. Used to figure out
+# what would be a good "index type" -- that is, a type that we can use to try
+# to look up matches with.
+class PromptTagTypeInfo
+  # We use a larger batch size here because prompts are smaller than signups
+  # (which can contain many prompts), so we can store more of them in memory at
+  # the same time.
+  def initialize(collection, batch_size = 300)
+    @collection = collection
+    @batch_size = batch_size
+
+    load_collection_info
+    initialize_data_tables
+  end
+
+  private
+
+  # Load information about the passed-in collection.
+  def load_collection_info
+    @total_prompts = @collection.prompts.count
+    @total_signups = @collection.signups.count
+    @settings = @collection.challenge.potential_match_settings
+    @required_types = @settings.required_types
+
+    @limit_first_tags = [10, @total_signups / 10].min
+    @limit_prompts_with_any = [10, @total_prompts / 10].max
+  end
+
+  # Set up @count_prompts_with_any and @first_tags_of_type with the correct
+  # defaults.
+  def initialize_data_tables
+    @count_prompts_with_any = {}
+    @first_tags_of_type = {}
+
+    @required_types.each do |type|
+      @count_prompts_with_any[type] = 0
+      @first_tags_of_type[type] = []
+    end
+  end
+
+  # A good type to use for indexing has a fair number of tags of that type (so
+  # that we aren't, e.g. trying to match prompts on fandom in a single-fandom
+  # exchange), and doesn't have very many people requesting "any" (so that we
+  # have a lot of constraints on who can match with whom).
+  def calculate_good_index_types
+    @good_index_types = @required_types.select do |type|
+      (@first_tags_of_type[type].size >= @limit_first_tags) &&
+        (@count_prompts_with_any[type] <= @limit_prompts_with_any)
```

Comment:
Not sure but talking out a thought here: it might work better to do this check purely on requests rather than on all prompts. Typically lots more people offer "any" rather than requesting it, especially with anything below fandom (character, ship, rating, etc), because there are many people who use as their only restriction on offering what they have the _knowledge_ to do, but their requests are much more specific. 

So eg, for a single-fandom (where the mod required 'fandom' anyway) or a few-fandom challenge (eg, an MCU challenge with say six Marvel movies as fandoms), I think looking at characters it's very likely to exceed the limit on 'any', and thus wind up creating the candidate batches on fandom as the top required tag, even though you are still better off in that scenario creating the candidate batches based on character rather than on fandom. 

That said, that challenge would probably have pretty dense matching no matter what unless there was some other constraining tag type below character. I'm fine leaving this alone and seeing what results we get from real-world data.